### DISCUSSION TITLE
Where does $data come from?

### DISCUSSION THREAD

**Discussion by s22-tech** [YOURLS/YOURLS#3260] [Q&A]
In the following example from https://github.com/SosthenG/yourls-notifier/, what is `$data` populated with and where does it come from?  It seems to be pulled out of thin air.  :grin:
```
yourls_add_filter('add_new_link', 'notify_new_link');
function notify_new_link($data) {
	...
}
```
Knowing how this works will help a lot in writing my own plugins.

Thanks,
Marc

**Comment by dgw** [comment]
It comes from the filter pipeline:

https://github.com/YOURLS/YOURLS/blob/a3cdf429c3a3cfb40f043fb8f4e3cdf09c0e9a94/includes/functions-shorturls.php#L162

`$return` is the first filter parameter, and becomes what that plugin calls `$data` when the filter's arguments are generated (by inspecting how many parameters the filter function expects):

https://github.com/YOURLS/YOURLS/blob/a3cdf429c3a3cfb40f043fb8f4e3cdf09c0e9a94/includes/functions-plugins.php#L224-L237

Finally, the [PHP docs for `call_user_func_array()`](https://www.php.net/manual/en/function.call-user-func-array.php):

> ```php
> call_user_func_array(callable $callback, array $args):
> ```
> Calls the callback given by the first parameter with the parameters in args.

**Reply by ozh** [reply]
See tutorial : https://github.com/YOURLS/YOURLS/wiki/How-to-make-Plugins
